internal/controllers/upgrade_handler.go

```diff
@@ -0,0 +1,164 @@
+package controllers
```

Comment:
I think we should rename this file to `provisioningrequest_cluster_upgrade.go` or something similar, but to have the controller using these functions prefixed. It would make it clear that it's used for provisioning.

Reply:
I already made quite a bit additional changes in this file. I will do the renaming in the subsequent PR.